[ScanDependency] Move binary module validation into scanner

cachemeifyoucan opened the pull request:
Improve swift dependency scanner by validating and selecting dependency module into scanner. This provides benefits that:
* Build system does not need to schedule interface compilation task if the candidate module is picked, it can just use the candidate module directly.
* There is no need for forwarding module in the explicit module build. Since the build system is coordinating the build, there is no need for the forwarding module in the module cache to avoid duplicated work,
* This also correctly supports all the module loading modes in the dependency scanner. This is achieved by only adding validate and up-to-date binary module as the candidate module for swift interface module dependency. This allows caching build to construct the correct dependency in the CAS. If there is a candidate module for the interface module, dependency scanner will return a binary module dependency in the dependency graph.

rdar://123711823

cachemeifyoucan committed:
[ScanDependency] Move binary module validation into scanner

Improve swift dependency scanner by validating and selecting dependency
module into scanner. This provides benefits that:
* Build system does not need to schedule interface compilation task if
  the candidate module is picked, it can just use the candidate module
  directly.
* There is no need for forwarding module in the explicit module build.
  Since the build system is coordinating the build, there is no need for
  the forwarding module in the module cache to avoid duplicated work,
* This also correctly supports all the module loading modes in the
  dependency scanner.

This is achieved by only adding validate and up-to-date binary module as
the candidate module for swift interface module dependency. This allows
caching build to construct the correct dependency in the CAS. If there
is a candidate module for the interface module, dependency scanner will
return a binary module dependency in the dependency graph.

The legacy behavior is mostly preserved with a hidden frontend flag
`-no-scanner-module-validation`, while the scanner output is mostly
interchangeable with new scanner behavior with `prefer-interface` module
loading mode except the candidate module will not be returned.

rdar://123711823

cachemeifyoucan commented:
This is not ready to land yet because I need to find way to fix the tests and clean up some code. Want to collect feedback for the implementation before doing all the extra work.

benlangmuir commented on lib/Frontend/ModuleInterfaceLoader.cpp:
```diff
@@ -1406,20 +1406,25 @@ ModuleInterfaceCheckerImpl::getCompiledModuleCandidatesForInterface(StringRef mo
+
+  // Add compiled module candidates only when they are non-empty and up-to-date.
+  // Since we are scanning interface, prefer the one in the cache.
+  if (!pair.second.empty() && validateModule(pair.second))
```
We should really make this a struct so the modules are named... If you don't want to change that in this PR you could use `std::tie(adjacentMod, prebuiltMod) = ` to destructure the pair.

benlangmuir commented on lib/Frontend/ModuleInterfaceLoader.cpp:
```diff
@@ -1406,20 +1406,25 @@ ModuleInterfaceCheckerImpl::getCompiledModuleCandidatesForInterface(StringRef mo
+  };
+
+  // Add compiled module candidates only when they are non-empty and up-to-date.
+  // Since we are scanning interface, prefer the one in the cache.
```
Is this a change in behaviour to prefer the prebuilt module?  I'm not sure I understand this

cachemeifyoucan replied:
Remove the comment and choose the order that is the same as original ordering.

benlangmuir commented on lib/Serialization/ScanningLoaders.cpp:
```diff
@@ -174,10 +174,6 @@ SwiftModuleScanner::scanInterfaceFile(Twine moduleInterfacePath,
           Args.push_back("-fno-implicit-module-maps");
         }
-        for (const auto &candidate : compiledCandidates) {
-          Args.push_back("-candidate-module-file");
```
Is `candidate_module_file` dead now? Maybe we can remove it from FrontendOptions.td

cachemeifyoucan replied:
I decide to preserve the flag and have an option to choose something that is very close to the original behavior. If everything goes well, we can remove the flag in the future.

benlangmuir commented on lib/DependencyScan/ScanDependencies.cpp:
```diff
@@ -192,14 +192,15 @@ updateModuleCacheKey(ModuleDependencyInfo &depInfo,
   if (moduleID.Kind == ModuleDependencyKind::SwiftPlaceholder)
     return llvm::Error::success();
 
+  auto resolvingDepInfo = cache.findKnownDependency(moduleID);
```
Probably `const auto &` to avoid a copy?  Applies to all other calls to this function as well

benlangmuir commented on lib/Frontend/ModuleInterfaceLoader.cpp:
```diff
@@ -1406,20 +1406,25 @@ ModuleInterfaceCheckerImpl::getCompiledModuleCandidatesForInterface(StringRef mo
+  if (!pair.second.empty() && validateModule(pair.second))
     results.push_back(pair.second);
+  else if (!pair.first.empty() && validateModule(pair.first))
+    results.push_back(pair.first);
```
Can `results` ever have more than one value now?  If not, can we simplify the result type?

cachemeifyoucan replied:
This patch just show the general structure. Ideally, I might switch the vector in to string in dependency but that is a lot of work for initial patch :)

benlangmuir reviewed (commented):
Approach looks reasonable to me, though I'm not an expert on this area of the code

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
Update: The legacy behavior is preserved with a hidden frontend flag
`-legacy-scanner-behavior`, while the output is mostly interchangeable
with new scanner behavior with `prefer-interface` module loading mode
except the candidate module will not be returned.

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
Rename the new flag name to `-no-scanner-module-validation` since it doesn't 100% preserve the old behavior but it is very very close, and delegate most of the module validation to swift-frontend and still generates forwarding module.

To clarify my previous comment, the behavior of the build should mostly stay the same (default to `prefer-serialized` loading mode) while the dependency scanning output from old compiler is very close to `prefer-interface` output except there is no candidate module reported.

cachemeifyoucan commented:
Need swift-driver change: https://github.com/apple/swift-driver/pull/1563 so the tests there are not broken because the change in scanner causes swift-driver to schedule less jobs.

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
ping

cachemeifyoucan committed:
[DependencyScanning] Handle testable dependencies correctly

Teach scanner to pick and choose binary modules correctly based on if it
is testable import or not. Some situations that scanner need to be
careful when testable is involved:

* When it is a regular import, it should not import binary modules that
  are built with -enable-testing, it should prefer interfaces if that is
  available.
* When testable import, it should only load binary module and it should
  make sure the internal imports from binary modules are actually
  required for testable import to work.

If a testable import only find a regular binary module, dependency
scanner currently will just preceed with such module and leave the
diagnostics to swift-frontend, because the alternative (failed to find
module) can be confusing to users.

rdar://125914165

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
@swift-ci please smoke test

artemcm commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -437,41 +425,60 @@ SerializedModuleLoaderBase::getImportsOfModule(
+      return std::make_error_code(std::errc::no_such_file_or_directory);
+    }
+
+    if (loadedModuleFile->isTestable() && !isTestableImport) {
```
Do we currently (before this change) have clients refuse to load `isTestable` binary modules from a non-`@testable` imports or is that a semantic change? 

In theory, a non-`@testable` `import` should be able to load such a module just fine, as it represents a super set of the same module built without `build-for-testing`.

cachemeifyoucan replied:
It can add in more dependency than needed which the user might or might not be able to satisfy.

Current explicit build will discover dependency from interface so it doesn't see those dependencies. I guess during forwarding module creation, it will see if it can satisfy all dependencies to see if it needs build from interface or create forwarding from testing module. That behavior is quite hard to support in this new model.

cachemeifyoucan commented:
@swift-ci please smoke test macOS platform
